src/main/java/teammates/logic/core/FeedbackSessionsLogic.java

```diff
@@ -1758,211 +1751,158 @@ private FeedbackSessionResultsBundle getFeedbackSessionResultsForUserWithParams(
-                            emailLastNameTable, emailTeamNameTable, sectionTeamNameTable,
-                            visibilityTable, responseStatus, roster, responseComments, true);
+        String section = params.get(PARAM_SECTION);
+        String questionId = params.get(PARAM_QUESTION_ID);
 
-            return results;
+        if (questionId != null) {
+            return getFeedbackSessionResultsForQuestionId(feedbackSessionName, courseId, userEmail, role, roster,
+                    session, relevantQuestions, allQuestions, isIncludeResponseStatus, section, questionId);
         }
 
         Map<String, FeedbackQuestionAttributes> allQuestionsMap = new HashMap<String, FeedbackQuestionAttributes>();
-        for (FeedbackQuestionAttributes qn : allQuestions) {
-            allQuestionsMap.put(qn.getId(), qn);
-        }
-        
+        putQuestionsIntoMap(allQuestions, allQuestionsMap);
+
         List<FeedbackResponseAttributes> allResponses = getAllResponses(feedbackSessionName, courseId,
-                params, section);
-        
-        boolean isComplete = params.get(PARAM_RANGE) == null;
-        
-        if (params.get(PARAM_RANGE) != null) {
-            long range = Long.parseLong(params.get(PARAM_RANGE));
-            if (allResponses.size() <= range) {
-                isComplete = true;
-            } else {
-                for (FeedbackQuestionAttributes qn : allQuestions) {
-                    relevantQuestions.put(qn.getId(), qn);
-                }
-                
-            }
+                                                                        params, section);
+
+        String rangeString = params.get(PARAM_RANGE);
+        boolean isComplete = rangeString == null || allResponses.size() <= Long.parseLong(rangeString);
+
+        if (!isComplete) {
+            putQuestionsIntoMap(allQuestions, relevantQuestions);
         }
         
-        responseStatus = section == null && isIncludeResponseStatus
+        FeedbackSessionResponseStatus responseStatus = section == null && isIncludeResponseStatus
                        ? getFeedbackSessionResponseStatus(session, roster, allQuestions)
                        : null;
 
-        StudentAttributes student = null;
-        Set<String> studentsEmailInTeam = new HashSet<String>();
-        if (role == UserRole.STUDENT) {
-            student = studentsLogic.getStudentForEmail(courseId, userEmail);
-            List<StudentAttributes> studentsInTeam = studentsLogic
-                    .getStudentsForTeam(student.team, courseId);
-            for (StudentAttributes teammates : studentsInTeam) {
-                studentsEmailInTeam.add(teammates.email);
-            }
-        }
-        
+        StudentAttributes student = getStudent(courseId, userEmail, role);
+
+        Set<String> teammateEmails = getTeammateEmails(courseId, student);
+
         InstructorAttributes instructor = null;
```

Comment:
@junhaoyap are you suggesting to refactor this "getting of instructor"? It does look very similar to what `getStudent(courseId, userEmail, role)` does. @samsontmr